### ISSUE TITLE
application can't create publisher repeatedly with previous one already destroyed

### ISSUE THREAD

**Issue by qclzdh** [eclipse-iceoryx/iceoryx#938]
## Required information

**Operating system:**
Ubuntu 20.04 LTS

**Compiler version:**
GCC 9.3.0

**Observed result or behaviour:**
test application reported some error log:
         **_[ Error ]: Request publisher received no valid publisher port from RouDi
         [Warning]: Service '10:Radar-test14:FrontLeft-test6:Object1:01:01:01:01:01:0' already in use by another process.
         [Warning]: ICEORYX error! POSH__RUNTIME_PUBLISHER_PORT_NOT_UNIQUE_**
Roudi log:
**_2021-10-08 15:27:33.443 [ Debug ]: Registered new application iox-cpp-publisher-helloworld
2021-10-08 15:27:33.444 [ Debug ]: Created new ApplicationPort for application iox-cpp-publisher-helloworld
2021-10-08 15:27:33.445 [ Debug ]: Created new PublisherPort for application iox-cpp-publisher-helloworld
2021-10-08 15:27:33.446 [ Debug ]: Created new PublisherPort for application iox-cpp-publisher-helloworld
2021-10-08 15:27:33.446 [Warning]: Process 'iox-cpp-publisher-helloworld' violates the communication policy by requesting a PublisherPort which is already used by 'iox-cpp-publisher-helloworld' with service '10:Radar-test14:FrontLeft-test6:Object1:01:01:01:01:01:0'.
2021-10-08 15:27:33.446 [Warning]: ICEORYX error! POSH__PORT_MANAGER_PUBLISHERPORT_NOT_UNIQUE
2021-10-08 15:27:33.446 [ Error ]: Could not create PublisherPort for application iox-cpp-publisher-helloworld
2021-10-08 15:27:33.541 [ Debug ]: Destroyed publisher port_**

**Expected result or behaviour:**
I expect the same publisher should be created successfully while previous one already destroyed.

**Conditions where it occurred / Performed steps:**
1. re-build whole iceoryx with "ONE_TO_MANY_ONLY=ON"
2. add following code in iceoryx_examples/icehello/iox_publisher_helloworld.cpp
    ` iox::capro::ServiceDescription serviceDescription{"Radar-test", "FrontLeft-test", "Object"};
      std::shared_ptr<iox::popo::Publisher<RadarObject>> p_publisher =
                  std::make_shared<iox::popo::Publisher<RadarObject>>(serviceDescription);
      p_publisher->stopOffer();
      p_publisher.reset();
      p_publisher = std::make_shared<iox::popo::Publisher<RadarObject>>(serviceDescription);`
3. build and run, roudi and iox-cpp-publisher-helloworld

**Comment by elBoberido** [comment]
@qclzdh currently we do not communicate the destruction of a port to RouDi to release the resources. The cleanup is done when the application stops and unregisters at RouDi or when an application crashes and RouDi does the cleanup. Up until now we did not have such a use case like you described because the publisher always lived as long as the application.

What is the use case for the scenario you described?

**Comment by elBoberido** [comment]
@qclzdh forget what I wrote about the release of the resource. It is released, but with a delay. In order to not block the destructor of the publisher only a flag is set and in the so called RouDi discovery loop the port is reclaimed by RouDi. So when you wait some time before the publisher is recreated it works.

```cpp
iox::capro::ServiceDescription serviceDescription{"Radar-test", "FrontLeft-test", "Object"};
auto p_publisher = std::make_shared<iox::popo::Publisher<RadarObject>>(serviceDescription);
p_publisher->stopOffer();
p_publisher.reset();
std::this_thread::sleep_for(std::chrono::milliseconds(2 * iox::roudi::DISCOVERY_INTERVAL.toMilliseconds()));
p_publisher = std::make_shared<iox::popo::Publisher<RadarObject>>(serviceDescription);
```

We could optimize this at the cost of slightly increasing the latency of creating new ports by checking the destruction flag for all the existing publisher ports and let RouDi reclaim a port which is marked for destruction before a new one is created

@budrus @elfenpiff @MatthiasKillat what do you think?

**Comment by budrus** [comment]
@elfenpiff @elBoberido This is a consequence of the delayed destruction in the discovery loop. I forgot the reasons why we did it like this and do not block the the d'tor and wait for RouDi to acknowledge the removal of the old port. Maybe to avoid the races with the discovery loop? I think we should fix this for the 2.0 release. Either like you propose by having a discovery (light) when creating a new port or by refactoring the cleanup

**Comment by budrus** [comment]
@qclzdh FYI, will be fixed with the upcoming 2.0 release

**Comment by qclzdh** [comment]
Thanks~
